core/runtime/common/trie_storage_provider_impl.cpp

```diff
@@ -53,96 +52,137 @@ namespace kagome::runtime {
+    if (child_it == child_batches_.end()) {
       SL_DEBUG(logger_,
-               "Creating new persistent batch for child storage {}",
+               "Creating new base batch for child storage {}",
                root_path.toHex());
-      OUTCOME_TRY(child_root_value, getCurrentBatch()->tryGet(root_path));
-      auto child_root_hash =
-          child_root_value
-              ? common::Hash256::fromSpan(*child_root_value).value()
-              : trie_serializer_->getEmptyRootHash();
-      OUTCOME_TRY(child_batch,
-                  trie_storage_->getPersistentBatchAt(child_root_hash));
-      child_batches_.emplace(root_path, std::move(child_batch));
+      OUTCOME_TRY(child_batch, base_batch_->createChildBatch(root_path));
+      // we check for duplicate batches in the if above
+      BOOST_ASSERT(child_batch.has_value());
+      child_batches_.emplace(root_path, child_batch.value());
+      return child_batch.value();
     }
-    SL_DEBUG(
-        logger_, "Fetching persistent batch for child storage {}", root_path);
-    return child_batches_.at(root_path);
+    SL_DEBUG(logger_, "Fetching batch for child storage {}", root_path);
+    return child_it->second;
   }
 
-  outcome::result<storage::trie::RootHash> TrieStorageProviderImpl::forceCommit(
-      StateVersion version) {
-    for (auto &[child, kv] : child_batches_) {
-      OUTCOME_TRY(root, kv->commit(version));
-      OUTCOME_TRY(current_batch_->put(child, common::BufferView{root}));
+  outcome::result<std::reference_wrapper<const storage::trie::TrieBatch>>
+  TrieStorageProviderImpl::getChildBatchAt(const common::Buffer &root_path) {
+    OUTCOME_TRY(batch_ptr, getOrCreateChildBatchAt(root_path));
+    return *batch_ptr;
+  }
+
+  outcome::result<std::reference_wrapper<storage::trie::TrieBatch>>
+  TrieStorageProviderImpl::getMutableChildBatchAt(
+      const common::Buffer &root_path) {
+    // if the batch for this child trie can be found in the current transaction,
+    // just return it
+    if (!transaction_stack_.empty()) {
+      if (auto it = transaction_stack_.back().child_batches.find(root_path);
+          it != transaction_stack_.back().child_batches.end()) {
+        return *it->second;
+      }
+      // if there are no open transactions, just a base batch is sufficient
+    } else {
+      OUTCOME_TRY(batch_ptr, getOrCreateChildBatchAt(root_path));
+      return *batch_ptr;
     }
-    if (persistent_batch_) {
-      return persistent_batch_->commit(version);
+    // otherwise we need to create a new overlay batch in the current
+    // transaction
+    OUTCOME_TRY(batch_ptr, getOrCreateChildBatchAt(root_path));
+
+    SL_DEBUG(logger_,
+             "Creating new overlay batch for child storage {}",
+             root_path.toHex());
+    auto child_batch =
+        std::make_shared<storage::trie::TopperTrieBatchImpl>(batch_ptr);
+    transaction_stack_.back().child_batches.emplace(root_path, child_batch);
+    return *child_batch;
+  }
```

Comment:
Looks like (unnecessary) duplicate of (complex) `getOrCreateChildBatchAt`.

Reply:
It's not a duplicate, it uses getOrCreateChildBatchAt to obtain the existing batch which is possibly on another transaction level, and then if the batch is not on the current transaction level it creates a batch on top of the obtained batch on the current transaction level.

Reply:
If it's necessary, then `getConstChildBatchAt` must return readonly batch, which errors on write attempt